src/impression.js

```diff
@@ -235,3 +240,66 @@ function applyResponse(win, response) {
     win.history.replaceState(null, '', newHref);
   }
 }
+
+/**
+ * Return a promise that whether appending extra url params to outgoing link is required.
+ * @param {!./service/ampdoc-impl.AmpDoc} ampdoc
+ * @return {!Promise<boolean>}
+ */
+export function shouldAppendExtraParams(ampdoc) {
+  return ampdoc.whenReady().then(() => {
+    return !!ampdoc.getBody().querySelector(
+        'amp-analytics[type=googleanalytics]');
+  });
+}
+
+/**
+ * Return the extra url params string that should be appended to outgoing link
+ * @param {!Window} win
+ * @param {!Element} target
+ * @return {string}
+ */
+export function getExtraParamsUrl(win, target) {
+  // Get an array with extra params that needs to append.
+  const url = parseUrl(win.location.href);
+  const params = parseQueryString(url.search);
+  const appendParams = [];
+  for (let i = 0; i < DEFAULT_APPEND_URL_PARAM.length; i++) {
+    const param = DEFAULT_APPEND_URL_PARAM[i];
+    if (typeof params[param] !== 'undefined') {
+      appendParams.push(param);
+    }
+  }
+
+  // Check if the param already exists
+  const additionalUrlParams = target.getAttribute('data-amp-addparams');
+  let href = target.href;
+  if (additionalUrlParams) {
+    href = addParamsToUrl(href, parseQueryString(additionalUrlParams));
+  }
+  const loc = parseUrl(href);
+  const existParams = parseQueryString(loc.search);
+  for (let i = appendParams.length - 1; i >= 0; i--) {
+    const param = appendParams[i];
+    if (typeof existParams[param] !== 'undefined') {
+      appendParams.splice(i, 1);
+    }
+  }
```

Comment:
This could all have been done in a single loop:

```js
let href = target.href;
const additionalUrlParams = target.getAttribute('data-amp-addparams');
if (additionalUrlParams) {
  href = addParamsToUrl(href, parseQueryString(additionalUrlParams));
}

const targetQuery = parseQueryString(parseUrl(href).search);
const windowQuery = parseQueryString(parseUrl(win.location.href).search);

const needed = DEFAULT_APPEND_URL_PARAM.filter(param => {
  return windowQuery[param] !== undefined && targetQuery[param] === undefined;
});
return getQueryParamUrl(needed);
```

Also, `typeof === 'undefined'` is unnecessary, just check `=== undefined`.